openbas-api/src/main/java/io/openbas/service/EndpointService.java

```diff
@@ -171,6 +174,90 @@ public void registerAgentEndpoint(AgentRegisterInput input) {
     }
   }
 
+  public List<Asset> syncAgentsEndpoints(
+      List<AgentRegisterInput> inputs, List<Agent> existingAgents) {
+    List<Agent> agentsToSave = new ArrayList<>();
+    List<Asset> endpointsToSave = new ArrayList<>();
+    Endpoint endpointToSave;
+    Agent agentToSave;
+    // Update agents/endpoints with external reference
+    Set<String> inputsExternalRefs =
+        inputs.stream().map(AgentRegisterInput::getExternalReference).collect(Collectors.toSet());
+    if (!inputsExternalRefs.isEmpty()) {
+      Set<Agent> agentsToUpdate =
+          existingAgents.stream()
+              .filter(agent -> inputsExternalRefs.contains(agent.getExternalReference()))
+              .collect(Collectors.toSet());
+      for (Agent agentToUpdate : agentsToUpdate) {
+        final AgentRegisterInput inputToSave =
```

Comment:
To avoid filtering again and again in the for loop, wouldn't a map created before the for loop be more efficient ?